File: ios/Sendro/Core/MessageCenter.swift
//
//  MessageCenter.swift
//  Sendro
//
//  The in-RAM inbox for PROTOCOL.md §11 text messages.
//
//  EPHEMERALITY CONTRACT — read before touching this file:
//  - Messages live in this array and nowhere else. No UserDefaults, no
//    Keychain, no file, no HistoryStore entry, no logging.
//  - `Message` is Decodable-only by design, so there is no way to serialize
//    one even by accident.
//  - `dismiss(id:)` drops the last reference; the memory goes with it.
//  - Quitting the app takes the whole inbox with it (no restore path).
//

import Foundation
import Combine

@MainActor
final class MessageCenter: ObservableObject {

    /// Oldest first; the UI shows the newest on top of the stack.
    @Published private(set) var inbox: [Message] = []

    /// §11: "A device's in-memory inbox holds at most 20 undelivered
    /// messages; pushing past that drops the oldest."
    static let capacity = 20

    nonisolated init() {}

    var hasMessages: Bool { !inbox.isEmpty }

    /// Ingest a batch straight off the outbox long poll. Deduped by
    /// messageId (delivery is at-most-once, but a resend must never
    /// double-stack a card).
    func receive(_ messages: [Message]) {
        guard !messages.isEmpty else { return }
        var changed = false
        for message in messages {
            guard !inbox.contains(where: { $0.messageId == message.messageId }) else { continue }
            inbox.append(message)
            changed = true
        }
        guard changed else { return }
        if inbox.count > Self.capacity {
            inbox.removeFirst(inbox.count - Self.capacity)
        }
    }

    /// Permanently forget one message.
    func dismiss(id: String) {
        inbox.removeAll { $0.messageId == id }
    }

    /// Permanently forget everything.
    func clear() {
        inbox.removeAll()
    }
}

File: android/app/src/main/java/com/sendro/android/core/MessageCenter.kt
package com.sendro.android.core

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * The in-RAM inbox for PROTOCOL.md §11 text messages.
 *
 * EPHEMERALITY CONTRACT — read before touching this file:
 *  - Messages live in this StateFlow and nowhere else. No DataStore, no
 *    SharedPreferences, no file, no HistoryStore entry, no Log call.
 *  - A message's text is never put in a notification body — [Notifier] posts
 *    "Sent you text" and the sender's name, nothing more.
 *  - [dismiss] drops the last reference; the memory goes with it.
 *  - Process death takes the whole inbox with it (there is no restore path,
 *    deliberately: §11.1 delivery is at-most-once).
 */
class MessageCenter {

    private val _inbox = MutableStateFlow<List<SendroMessage>>(emptyList())

    /** Oldest first; the UI stacks the newest on top. */
    val inbox: StateFlow<List<SendroMessage>> = _inbox.asStateFlow()

    /**
     * Ingest a batch straight off the outbox long poll. Deduped by messageId:
     * delivery is at-most-once, but a resend must never double-stack a card.
     */
    @Synchronized
    fun receive(messages: List<SendroMessage>) {
        if (messages.isEmpty()) return
        val current = _inbox.value
        val known = current.mapTo(HashSet()) { it.messageId }
        val fresh = messages.filter { known.add(it.messageId) }
        if (fresh.isEmpty()) return
        // §11: "at most 20 undelivered messages; pushing past that drops the
        // oldest."
        _inbox.value = (current + fresh).takeLast(CAPACITY)
    }

    /** Permanently forget one message. */
    @Synchronized
    fun dismiss(messageId: String) {
        _inbox.value = _inbox.value.filterNot { it.messageId == messageId }
    }

    /** Permanently forget everything. */
    @Synchronized
    fun clear() {
        _inbox.value = emptyList()
    }

    companion object {
        const val CAPACITY = 20
    }
}
